lib/toolbox.js

```diff
@@ -62,10 +62,14 @@ pkijs.setEngine(
 */
 function derToRaw(signature) {
 	const rStart = signature[4] === 0 ? 5 : 4;
-	const rEnd = rStart + 32;
+	const rEnd = 4 + signature[3];
```

Comment:
I'm not sure it makes sense to hardcode this, given that there is a check above. The idea of the line above is exactly to check for a null in 3rd position, so where exactly is your difference? Please expand the evaluation and keep using the existing parts carefully. The slicing below probably should also only trigger in those cases.

Reply:
The length of r is not always 32 bytes. The check above (line 64) handles the case where the first byte is a zero because the first byte of r is a signed negative number. This ensures the value is a positive number.

However, ASN.1/DER tries to minimize the number (removing leading zeros). So if you have an r that is `000000abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123` it will be minimized to `00abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123` (leading 00 because ab is negative).
When you have this as r, the length field will be 30 instead of 32. So if you use the hardcoded 32 byte value, you are grabbing values that are part of s, which will make the signature invalid.

Raw/IEEE P1363 requires the r/s values to be non-minimized unsigned values (always 32 bytes), so we need to add back the 00 padding to the number to make it 32 bytes when we return the value.

Reply:
Understood, however your code always runs and executes for all those cases. If you know that in some cases some padding or stripping is required, it'd be best to check for exactly those situations and apply it only then.

Reply:
I'm not sure I follow. Are you suggesting that for only the padding part? For example would this be closer to what you are wanting?
```
function derToRaw(signature) {
	const rStart = signature[4] === 0 ? 5 : 4;
	const rEnd = 4 + signature[3];
	const sStart = signature[rEnd + 2] === 0 ? rEnd + 3 : rEnd + 2;
	let r = signature.slice(rStart, rEnd);
	let s = signature.slice(sStart);
	if (r.length !== 32) {
		r = Array(32).fill(0).concat(...r).slice(r.length);
	}
	if (s.length !== 32) {
		s = Array(32).fill(0).concat(...s).slice(s.length);
	}
	return new Uint8Array([
		...r,
		...s,
	]);
}
```

Or are you suggesting to also check the length byte and then if it is 32/33 use 32 instead of the length?
```
function derToRaw(signature) {
	const rStart = signature[4] === 0 ? 5 : 4;
	let rEnd = rStart + 32;
	if (rEnd > signature[3] + 4) {
		rEnd = signature[3] + 4;
	}
	const sStart = signature[rEnd + 2] === 0 ? rEnd + 3 : rEnd + 2;
	let r = signature.slice(rStart, rEnd);
	let s = signature.slice(sStart);
	if (r.length !== 32) {
		r = Array(32).fill(0).concat(...r).slice(r.length);
	}
	if (s.length !== 32) {
		s = Array(32).fill(0).concat(...s).slice(s.length);
	}
	return new Uint8Array([
		...r,
		...s,
	]);
}
```

Reply:
If the first one is ok... This simplifies the function a bit, since the slice part will handle adding and removing any leading padding:
```
function derToRaw(signature) {
	const rStart = 4;
	let rEnd = rStart + signature[3];
	const sStart = rEnd + 2;
	let r = signature.slice(rStart, rEnd);
	let s = signature.slice(sStart);
	if (r.length !== 32) {
		r = Array(32).fill(0).concat(...r).slice(r.length);
	}
	if (s.length !== 32) {
		s = Array(32).fill(0).concat(...s).slice(s.length);
	}
	return new Uint8Array([
		...r,
		...s,
	]);
}
```

Reply:
Thank you for the proposals, this looks more like going the right direction - however the code is quite repetitive and you are checking for unequal instead of smaller. However in your example you are using `+ signature[3]` - are we sure this cannot be a letter? In your example payload it is.

From my understanding and your current code and examples, we basically just need to pad the signature to 64 characters with leading zeros, isn't it? Some encodings remove it at the beginning, we add it back on? If my understanding is correct, can we not just use `signature = signature.padStart(64, "0")`?

Reply:
> you are checking for unequal instead of smaller

Depending on which version you are referring to, unequal is intended. In the final version, unequal handles the case where the length is 33 (zero padded to make the number positive), and the cases where the length is less than 32.
In the first two, unequal is still accurate, but since the length will never be above 32, less than works as well.

> However in your example you are using + signature[3] - are we sure this cannot be a letter?

What do you mean? Are you saying that the value will be invalid?
My example is just the r or s, a full signature would look like this:
```
> new Uint8Array(Buffer.from('MEMCHhCs-kZTCokgrPfb1CaKEznJjqVisSBzMAqv6S24AQIhAOBeIIWXFgOJwA-39dKfzcuG6woJ03tiR0N2ME9Lp206', 'base64url'))
Uint8Array(69) [
   48,  67,   2,  30,  16, 172, 250,  70,  83,  10, 137,  32,
  172, 247, 219, 212,  38, 138,  19,  57, 201, 142, 165,  98,
  177,  32, 115,  48,  10, 175, 233,  45, 184,   1,   2,  33,
    0, 224,  94,  32, 133, 151,  22,   3, 137, 192,  15, 183,
  245, 210, 159, 205, 203, 134, 235,  10,   9, 211, 123,  98,
   71,  67, 118,  48,  79,  75, 167, 109,  58
]
```
For a valid signature, the fourth byte (sig[3]) will always be the length of r, and the `sig[3] + 4 + 1` byte will always be the length of s.

If the length byte is not accurate, then the signature is not valid, so converting it to something that is invalid should not be an issue. We could be defensive and check that the length is not larger than the size of the buffer. However, given we are not directly reading the array (outside of byte 4), but instead using slice, slice will handle the range checks for us.

> From my understanding and your current code and examples, we basically just need to pad the signature to 64 characters with leading zeros, isn't it? Some encodings remove it at the beginning, we add it back on? If my understanding is correct, can we not just use signature = signature.padStart(64, "0")?

No. We need r and s to each be 32 byte numbers. In the case where both r and s are less than 32 bytes in the ASN.1 signature, if you combined the smaller numbers first then pad the final number, the signature would not be valid.

For example, given this ASN.1 signature:
```
Uint8Array(68) [
   48,  66,   2,  31,  21, 133, 215, 143,   9, 144,  85, 124,
  235, 243, 115,  83, 224,  15,  63, 236, 161,  86, 229,   7,
   18,  37,  90, 180, 119,  94,  15, 162, 186,   3,  64,   2,
   31,   5, 172,  49, 100,  40, 223, 122,  18,   6,   2, 119,
  105,  37,  94,  99,  87,  98, 236,  31, 211,  83, 129, 131,
    1,  61,  37, 221,  11, 111,  61, 136
]
```
Minimized R is:
```
Uint8Array(31) [
   21, 133, 215, 143,   9, 144,  85, 124,
  235, 243, 115,  83, 224,  15,  63, 236, 161,  86, 229,   7,
   18,  37,  90, 180, 119,  94,  15, 162, 186,   3,  64
]
```
R is:
```
Uint8Array(32) [
     0,  21, 133, 215, 143,   9, 144,  85, 124,
  235, 243, 115,  83, 224,  15,  63, 236, 161,  86, 229,   7,
   18,  37,  90, 180, 119,  94,  15, 162, 186,   3,  64
]
```

Minimized S is:
```
Uint8Array(31) [
    5, 172,  49, 100,  40, 223, 122,  18,   6,   2, 119,
  105,  37,  94,  99,  87,  98, 236,  31, 211,  83, 129, 131,
    1,  61,  37, 221,  11, 111,  61, 136
]
```
S is:
```
Uint8Array(32) [
    0,    5, 172,  49, 100,  40, 223, 122,  18,   6,   2, 119,
  105,  37,  94,  99,  87,  98, 236,  31, 211,  83, 129, 131,
    1,  61,  37, 221,  11, 111,  61, 136
]
```

If you combine then pad, your signature is:
```
Uint8Array(64) [
     0,    0,  21, 133, 215, 143,   9, 144,  85, 124,
  235, 243, 115,  83, 224,  15,  63, 236, 161,  86, 229,   7,
   18,  37,  90, 180, 119,  94,  15, 162, 186,   3,  64,
    5, 172,  49, 100,  40, 223, 122,  18,   6,   2, 119,
  105,  37,  94,  99,  87,  98, 236,  31, 211,  83, 129, 131,
    1,  61,  37, 221,  11, 111,  61, 136
]
```
If you then parse r and s, you get r:
```
Uint8Array(64) [
     0,    0,  21, 133, 215, 143,   9, 144,  85, 124,
  235, 243, 115,  83, 224,  15,  63, 236, 161,  86, 229,   7,
   18,  37,  90, 180, 119,  94,  15, 162, 186,   3
]
```
And s:
```
Uint8Array(64) [
  64, 5, 172,  49, 100,  40, 223, 122,  18,   6,   2, 119,
  105,  37,  94,  99,  87,  98, 236,  31, 211,  83, 129, 131,
    1,  61,  37, 221,  11, 111,  61, 136
]
```
Which is not correct.

> however the code is quite repetitive

I'm open to suggests to avoid repetitiveness. What about:
```
function extractBigNum(fullArray, start, end, expectedLength) {
	const num = fullArray.slice(start, end);
	return Array(expectedLength).fill(0).concat(...num).slice(num.length);
}

/*
    Convert signature from DER to raw
    Expects Uint8Array
*/
function derToRaw(signature) {
	const rStart = 4;
	const rEnd = rStart + signature[3];
	const sStart = rEnd + 2;
	return new Uint8Array([
		...extractBigNum(signature, rStart, rEnd, 32),
		...extractBigNum(signature, sStart, signature.length, 32),
	]);
}
```

Reply:
Thanks for the detailed suggestion! I was under the wrong assumption that signature was a string, but you're right, it's not. The new code looks great - we can close this part 👍